ngraph/core/reference/include/ngraph/runtime/reference/psroi_pooling.hpp

```diff
@@ -0,0 +1,204 @@
+//*****************************************************************************
+// Copyright 2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <cmath>
+#include <string>
+
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            enum PSROIPoolingMode
+            {
+                AVG,
+                BILINEAR
+            };
+            template <typename T>
+            void psroi_pooling(const T* input,
+                               const Shape& input_shape,
+                               const T* rois,
+                               const Shape& rois_shape,
+                               T* output,
+                               const Shape& output_shape,
+                               const std::string& mode_str,
+                               float spatial_scale,
+                               int spatial_bins_x,
+                               int spatial_bins_y)
+            {
+                PSROIPoolingMode mode;
+                if (mode_str == "average")
+                {
+                    mode = AVG;
+                }
+                else if (mode_str == "bilinear")
+                {
+                    mode = BILINEAR;
+                }
+                else
+                {
+                    NGRAPH_CHECK(false, "Invalid PS ROI pooling mode");
+                }
+                int channels_in = input_shape[1];
```

Comment:
in general dimension size can be 64 bits (and it is unsigned value), but size of int is 32 bits, it means that you can face with Integer overflow.
Please check all other places.